### ISSUE TITLE
[macOS Silicon] Table and Tree item height gets smaller when setting the font

### ISSUE THREAD

**Issue by Phillipus** [eclipse-platform/eclipse.platform.swt#677]
This occurs on  Mac Silicon (aarch64) only, presumably because SWT on that platform uses a later Apple SDK. I've tested on Mac Mini M1 with Monterey and Ventura.

Setting a table or tree font to the same (or different) font reduces the item height.

1. Create a simple table with some rows
2. The item height of the rows is 25:
<img width="612" alt="Before" src="https://github.com/eclipse-platform/eclipse.platform.swt/assets/600504/8e3583c5-7121-4e52-8666-b1346de46359">

3. Set the table's font to its current font (`table.setFont(table.getFont())`)
4. The item height of the rows is now 18:
<img width="612" alt="After" src="https://github.com/eclipse-platform/eclipse.platform.swt/assets/600504/3a293388-68a5-430a-945d-cf1973f90755">

£xpected behaviour: should use the height of the target SDK.  On aarch64 this is the 25 pixel item height. On Intel it is 18.

**Comment by Phillipus** [comment]
Snippet to reproduce:

```java
import static org.eclipse.swt.events.SelectionListener.widgetSelectedAdapter;

import org.eclipse.jface.layout.GridDataFactory;
import org.eclipse.swt.SWT;
import org.eclipse.swt.layout.GridData;
import org.eclipse.swt.layout.GridLayout;
import org.eclipse.swt.widgets.Button;
import org.eclipse.swt.widgets.Display;
import org.eclipse.swt.widgets.Label;
import org.eclipse.swt.widgets.Shell;
import org.eclipse.swt.widgets.Table;
import org.eclipse.swt.widgets.TableItem;

public class TableFont {

    public static void main(String[] args) {
        Display display = new Display();
        Shell shell = new Shell(display);
        shell.setText("Table Font");
        shell.setLayout(new GridLayout());
        
        Table table = new Table(shell, SWT.BORDER | SWT.V_SCROLL | SWT.H_SCROLL);
        GridDataFactory.create(GridData.FILL_HORIZONTAL).applyTo(table);
        
        for(int i = 0; i < 11; i++) {
            TableItem item = new TableItem(table, 0);
            item.setText("Item " + i);
        }
        
        Label label1 = new Label(shell, SWT.NONE);
        Label label2 = new Label(shell, SWT.NONE);
        updateLabels(table, label1, label2);
        
        Button button = new Button(shell, SWT.PUSH);
        button.setText("Set font");
        button.addSelectionListener(widgetSelectedAdapter(e -> {
            // Set table font to the existing table font
            table.setFont(table.getFont());
            updateLabels(table, label1, label2);
        }));
        
        shell.setSize(500, 450);
        shell.open();
        
        while(!shell.isDisposed()) {
            if(!display.readAndDispatch()) display.sleep();
        }
        
        display.dispose();
    }

    private static void updateLabels(Table table, Label label1, Label label2) {
        label1.setText("Table font: " + table.getFont().getFontData()[0]);
        label2.setText("Table item height: " + table.getItemHeight());
    }
}
```

**Comment by elsazac** [comment]
@Phillipus 
Can you also add what should be the expected behaviour?

**Comment by Phillipus** [comment]
>  Can you also add what should be the expected behaviour?

Added to OP.

**Comment by elsazac** [comment]
I am investigating on this issue and these are my findings.
https://github.com/eclipse-platform/eclipse.platform.swt/blob/1a939af1dfb0d050f91e5357887fa068f336da23/bundles/org.eclipse.swt/Eclipse%20SWT/cocoa/org/eclipse/swt/widgets/Table.java#L2832-L2833

If I change the above condition either to one of the below , we see the expected behaviour but not sure about the exact purpose of `set`. Does someone has any idea on when exactly `set` needs to be evaluated?
```
if (widget.rowHeight () < height) {
		widget.setRowHeight (height);
`````

```
if (set && widget.rowHeight () < height) {
		widget.setRowHeight (height);
```

**Comment by Phillipus** [comment]
Hi, I tested the patch. If there are no items the heights are too small:

<img width="612" alt="s" src="https://github.com/eclipse-platform/eclipse.platform.swt/assets/600504/66611bb4-1fa7-4633-a8a2-8dc7e64ada1b">

```java
package com.archimatetool.editor;
import org.eclipse.jface.layout.GridDataFactory;
import org.eclipse.swt.SWT;
import org.eclipse.swt.layout.GridData;
import org.eclipse.swt.layout.GridLayout;
import org.eclipse.swt.widgets.Display;
import org.eclipse.swt.widgets.Shell;
import org.eclipse.swt.widgets.Table;
import org.eclipse.swt.widgets.Tree;

public class FontTest {

    public static void main(String[] args) {
        Display display = new Display();
        Shell shell = new Shell(display);
        shell.setText("Tree & Table Font");
        shell.setLayout(new GridLayout());
        
        Tree tree = new Tree(shell, SWT.BORDER | SWT.V_SCROLL | SWT.H_SCROLL);
        GridDataFactory.create(GridData.FILL_BOTH).applyTo(tree);
        tree.setLinesVisible(true);
        
        Table table = new Table(shell, SWT.BORDER | SWT.V_SCROLL | SWT.H_SCROLL);
        GridDataFactory.create(GridData.FILL_BOTH).applyTo(table);
        table.setLinesVisible(true);
        
        shell.setSize(500, 450);
        shell.open();
        
        tree.setFont(tree.getFont());
        table.setFont(table.getFont());
        
        while(!shell.isDisposed()) {
            if(!display.readAndDispatch()) display.sleep();
        }
        
        display.dispose();
    }
}
```

**Comment by Phillipus** [comment]
The problem also occurs for the SWT `List` widget:

```java
import static org.eclipse.swt.events.SelectionListener.widgetSelectedAdapter;

import org.eclipse.jface.layout.GridDataFactory;
import org.eclipse.swt.SWT;
import org.eclipse.swt.layout.GridData;
import org.eclipse.swt.layout.GridLayout;
import org.eclipse.swt.widgets.Button;
import org.eclipse.swt.widgets.Display;
import org.eclipse.swt.widgets.Label;
import org.eclipse.swt.widgets.List;
import org.eclipse.swt.widgets.Shell;

public class ListRowHeight {

    public static void main(String[] args) {
        Display display = new Display();
        Shell shell = new Shell(display);
        shell.setText("List Row Height");
        shell.setLayout(new GridLayout());
        
        List list = new List(shell, SWT.BORDER | SWT.V_SCROLL | SWT.H_SCROLL);
        GridDataFactory.create(GridData.FILL_HORIZONTAL).applyTo(list);
        
        String[] items = new String[10];
        for(int i = 0; i < 10; i++) {
            items[i] = "Item " + i;
        }
        list.setItems(items);
        
        Label label1 = new Label(shell, SWT.NONE);
        Label label2 = new Label(shell, SWT.NONE);
        updateLabels(list, label1, label2);
        
        Button button = new Button(shell, SWT.PUSH);
        button.setText("Set font");
        button.addSelectionListener(widgetSelectedAdapter(e -> {
            // Set list font to the existing table font
            list.setFont(list.getFont());
            updateLabels(list, label1, label2);
        }));
        
        shell.setSize(500, 450);
        shell.open();
        
        while(!shell.isDisposed()) {
            if(!display.readAndDispatch()) display.sleep();
        }
        
        display.dispose();
    }

    private static void updateLabels(List list, Label label1, Label label2) {
        label1.setText("List font: " + list.getFont().getFontData()[0]);
        label2.setText("List item height: " + list.getItemHeight());
    }
}
```

**Comment by Phillipus** [comment]
`Table` and `Tree` row heights are initially set to 25 on Mac aarch64 and then set to 18 when resetting the font. With a `List` widget the row height is initially 24 and then 17 when resetting the font.

I have a doubt about whether the initial row height is wrong and should be as it is on intel Mac. But there are these earlier bug reports:

https://bugs.eclipse.org/bugs/show_bug.cgi?id=575696

https://bugs.eclipse.org/bugs/show_bug.cgi?id=574932

and this comment from @sravanlakkimsetti:

https://bugs.eclipse.org/bugs/show_bug.cgi?id=574932#c2

> This is because eclipse launcher SDK version
> On x86_64 SDK used is 10.14
> On aarch64 SDK used is 11
> But the way forward is to use latest SDK and update SWT to MacOS 11 SDK.

So I guess the higher row height values are correct for aarch64?

**Comment by elsazac** [comment]
@Phillipus 

I have verified some tables in Mac and I can see the table height in Mac being aligned with table in the snippet(with height 25)

<img width="971" alt="Screenshot 2023-12-06 at 1 11 26 PM" src="https://github.com/eclipse-platform/eclipse.platform.swt/assets/122080218/19052d76-a089-42bc-9728-49ed49734640">

I have gone through the bugzilla links and I see that the increased item height is the new native behaviour on BigSur. row height is 24pt in BigSur and 17pt in macOS 10.15.
https://bugs.eclipse.org/bugs/show_bug.cgi?id=574932#c40

Also can you please try some tables in x86_64?

**Comment by tmssngr** [comment]
What is the expected result of the last commit 611accb8? For our applications users complain that the line height is much higher with the latest SWT build than with previous versions. However, I don't see any change with this commit.

**Comment by merks** [comment]
I assume that you need to ensure that this returns false:
```
Boolean.parseBoolean(System.getProperty("org.eclipse.swt.internal.cocoa.useNativeItemHeight", "true"));
```

**Comment by kohlschuetter** [comment]
@tmssngr For apps that use `-Dorg.eclipse.swt.internal.carbon.smallFonts` (like Eclipse IDE), the change in https://github.com/eclipse-platform/eclipse.platform.swt/commit/611accb844fb525efb3727f7fe8e2e7a3472debf brings back the old behavior, as an immediate remedy to the problem without the need of further configuration.

For apps that don't specify `smallFonts`, the appearance depends on their macOS version. Newer versions of macOS have a taller native height that matters here. To consistently bring back the old behavior, specify `-Dorg.eclipse.swt.internal.cocoa.useNativeItemHeight=false` when launching your app.

**Comment by tmssngr** [comment]
I'm not sure I fully understand all. Older versions of SWT (e.g. used in SmartGit 23.1.2) show table and tree rows quite narrow on my M1 Mac Mini with macOS 14.4. Newer versions of SWT (e.g. used in SmartGit 24.1 preview) show a much larger row height. We don't have the mentioned properties set. Do you mean, that the larger row height in newer SWT builds is now the correct behavior for macOS 14.4, and if a user wants to have a narrower row-height, he needs to set the mentioned properties?

**Comment by kohlschuetter** [comment]
@tmssngr Yes, I'm afraid that's the case, but setting a property will do the trick.

Anyhow, I'm just a Eclipse IDE user whose eyes couldn't bear looking at these oversized list items, so I'm glad I could contribute to resolving this issue.

**Comment by Phillipus** [comment]
> Older versions of SWT (e.g. used in SmartGit 23.1.2) show table and tree rows quite narrow on my M1 Mac Mini with macOS 14.4.

Take a look at SmartGit 23.1.2 Settings tree. You'll see bigger row heights.

There always was bigger row heights since the later macOS versions. Hwoever, 
when a call to `setItemHeight` is made (by setting the control's font or image or something else), the row heights shrink,  as the snippet here demonstrates. The commit of @kohlschuetter corrects this anomaly.

**Comment by tmssngr** [comment]
@Phillipus Ah, so there was even another difference between plain and owner-drawn trees.
On MacOS we were setting the display's system font (a slightly different one) to have all digits the same width. So it looks like our problem of the increased row-heights is a slightly different problem than this one covered in this ticket. Sorry for the noise.

**Comment by Phillipus** [comment]
@tmssngr

It depends on how a Tree/Table/List is used.

As the snippet in this issue shows, when a Tree/Table/List control is first created it has the native row height (larger on later macOS). Thereafter, a change to the control's font/image/whatever triggers a re-calculation of the row height. This is where the row height shrinks because it doesn't take into account the initial native row height when calculating the row height based upon the current font.

Eclipse uses the `-Dorg.eclipse.swt.internal.carbon.smallFonts` property and this triggers the recalculation of row heights immediately after the control is created. This is why Eclipse users always saw smaller row heights.

Now that the recalculation anomaly has been fixed there is consistency between the initial row heights and the re-calculated ones, so no more unexpected row height shrinkage. However, this means bigger row heights by default because that's how they are initially created.

I think SmartGit Preview is seeing bigger row heights because of the erroneous padding added in [this commit](https://github.com/eclipse-platform/eclipse.platform.swt/commit/0032340b9b647db3ef6776d6b2e481397bd74b8d).

You could try using the latest SWT build with the new fix in and add the `-Dorg.eclipse.swt.internal.cocoa.useNativeItemHeight=false` setting to SmartGit's `plist.info` file (I think that's where you set these properties?).

**Comment by tmssngr** [comment]
Thanks, @Phillipus, this helped to fix it. :)

**Comment by odrotbohm** [comment]
I might be the odd one out here, but is there a way to get both small fonts, but the bigger spacing back after the fix? 😇 I see my ini file having `-Dorg.eclipse.swt.internal.carbon.smallFonts` set. I've tried adding `-Dorg.eclipse.swt.internal.cocoa.useNativeItemHeight=true` (I also tried `false`) to get the bigger spacing back, but don't see any effect.

**Comment by Phillipus** [comment]
> is there a way to get both small fonts, but the bigger spacing back after the fix?

No, because the native item height is smaller when small fonts are set.

**Comment by odrotbohm** [comment]
That's a pity, but thanks! 😕

**Comment by marcleidner** [comment]
Issue is fixed in 2024-06 (4.32.0), for me at least.
